tokio-postgres/src/passfile/mod.rs

```diff
@@ -0,0 +1,199 @@
+//! Support for reading password files.
+//!
+//! Requires the `runtime` Cargo feature.
+
+#[cfg(unix)]
+use std::os::unix::{ffi::OsStrExt, fs::MetadataExt};
+use std::path::Path;
+
+use log::warn;
+use tokio::fs::{self, File};
+use tokio::io::BufReader;
+
+use crate::config::{Config, Host};
+
+#[cfg(test)]
+mod test;
+
+/// The data needed to search for a matching passfile entry.
+struct PassfileKey<'a> {
+    hostname: &'a [u8],
+    port: Vec<u8>,
+    dbname: &'a [u8],
+    user: &'a [u8],
+}
+
+impl<'a> PassfileKey<'a> {
+    fn new(host: &'a Host, port: u16, dbname: Option<&'a str>, user: &'a str) -> PassfileKey<'a> {
+        let hostname = match host {
+            Host::Tcp(s) => s.as_bytes(),
+            #[cfg(unix)]
+            // libpq translates DEFAULT_PGSOCKET_DIR to 'localhost' here, but we can't do the same that because we don't
+            // know what DEFAULT_PGSOCKET_DIR is.
+            Host::Unix(pathbuf) => pathbuf.as_os_str().as_bytes(),
+        };
+        let port_string = format!("{}", port).into_bytes();
+        // This default is applied by the server, rather than our caller, so we have to apply it here too.
+        let dbname = dbname.unwrap_or(user);
+        PassfileKey {
+            hostname,
+            port: port_string,
+            dbname: dbname.as_bytes(),
+            user: user.as_bytes(),
+        }
+    }
+}
+
+/// The data from a single passfile line.
+struct PassfileEntry {
+    hostname: Vec<u8>,
+    port: Vec<u8>,
+    dbname: Vec<u8>,
+    user: Vec<u8>,
+    password: Vec<u8>,
+}
```

Comment:
You might consider a stronger type representation here, e.g. `u16` for the port, and maybe `Option` to represent the stars.

Maybe this would be counterproductive, so it's just an idea. I'd value compatibility with libpq behavior over type cleanliness.

Reply:
Thanks, I've introduced a `PassfileField` type to represent wildcards cleanly. I don't think representing the port as an integer would be helpful (in particular, we don't want to allow "05432" to match port 5432).